app/upload/views.py

```diff
@@ -12,16 +16,26 @@
 
 @upload_blueprint.route('/services/<uuid:service_id>/documents', methods=['POST'])
 def upload_document(service_id):
-    if 'document' not in request.files:
-        return jsonify(error='No document upload'), 400
+    no_document_error = jsonify(error='No document upload'), 400
+
+    if request.is_json:
+        if 'document' not in request.json:
+            return no_document_error
+        raw_content = b64decode(request.json['document'])
+        if len(raw_content) > current_app.config['MAX_CONTENT_LENGTH']:
```

Comment:
Note that we have to perform this content length check manually for the case of a file in the JSON body - Flask's built in content length checking does not check POST bodies.

I believe that this is slightly different to how flask's length checking works: with this solution, we have to parse the request body (and decode the document string) before determining that the file is too large.  I believe that Flask's built in solution will evaluate much sooner in the request/response cycle and thus avoid parsing the request to this point.

It would be good to get some comments on whether this approach is OK.

Reply:
i think this is fine. this check is just a rule of thumb designed to prevent people uploading documents that are going to be a pain for end users to download on bad connections.

We'll catch really obscenely big payloads much earlier (we limit to 5mb at our boundary nginx instances) so i'm not worried about the risk of someone uploading a huge file with the intent of DOSing us when we try and json decode it. We need to read the entire blob anyway to parse the json, so this is fine 👍.